### PR TITLE
Add support for storing game information in wesnoth's mysql database.

### PR DESCRIPTION
**Pull request by Pentarctagon**
This requires three more tables to be added to any database with the forum user handler enabled, the structures of which are defined in the wesnothd man page:
* `db_game_info_table` - stores information about each game.
* `db_game_player_info_table` - stores information about each player in the game.
* `db_game_modification_info_table` - stores information about any modifications that are enabled for the game.

### CODE REVIEW

**Comment by ProditorMagnus** [comment]
Do you want comments about data model?

**Comment by Pentarctagon** [comment]
Sure.

**Comment by ProditorMagnus** [comment]
**Game info**
>   GAME_ID            SMALLINT UNSIGNED NOT NULL,
>   primary key (INSTANCE_UUID, GAME_ID)

Small int is not enough for stable series. Is the GAME_ID primary key just as index as it is not unique?

>   PLAYER_COUNT       TINYINT UNSIGNED,
>  SIDE_COUNT         TINYINT UNSIGNED,

What is the purpose of these?

>  REPLAY_NAME        VARCHAR(255),

How it differs from GAME_NAME?

**Player info**
How you handle cases where someone changes control of sides during game?

>   GAME_ID       SMALLINT UNSIGNED NOT NULL,

Game id is not unique, so it is not enough to identify single row in game table.

**Modification info**
I have not heard anything about core modifications, and for addons you would need to consider addon version and options.

**Comment by Pentarctagon** [comment]
> Small int is not enough for stable series. Is the GAME_ID primary key just as index as it is not unique?

The GAME_ID is the sequential ID that wesnothd creates.  It isn't unique, since every time wesnothd restarts, wesnothd restarts at 1.  So that is why it is an UNSIGNED (wesnothd, afaik, will never create an id below 0) and SMALLINT (it seemed unlikely that there would be 65k+ matches between restarts of wesnothd).

INSTANCE_UUID is the value returned by mysql's `UUID()` function, which wesnothd would query at startup.  That is why the composite primary key is of INSTANCE_UUID and GAME_ID - it's the unique ID retrieved per instance plus the sequential game id of that particular instance.

> What is the purpose of these?

My thought for those was to show cases where for example Isar's has 4 sides, but each person controls 2 sides.  Though I suppose that's somewhat redundant, given `db_game_player_info_table` exists.

> How it differs from GAME_NAME?

It would be the path to the file, as saved at replays.wesnoth.org.  My thought being that, in the future, it would allow people either in-client or via the Wesnoth site to search for games and be able to immediately download the replays of the games they're looking for.

> How you handle cases where someone changes control of sides during game?

That case isn't handled, currently.  The side control change could be updated onto the STATUS column, I suppose, though it makes it more difficult to create a primary key - just adding user_id wouldn't be enough, since it wouldn't cover the case of the side changing owners back and forth multiple times, perhaps due to poor internet connection.

> Game id is not unique, so it is not enough to identify single row in game table.

See first response.

> I have not heard anything about core modifications, and for addons you would need to consider addon version and options.

I agree that would be useful to know, but I'm not sure where that information is available as part of the game instance information(or if it is available).  The same would apply to non-mainline scenarios and eras as well.

**Comment by Pentarctagon** [comment]
Removed PLAYER_COUNT and SIDE_COUNT as they're redundant.

**Comment by GregoryLundberg** [comment]
"seems unlikely" --> "famous last words"

A little time thinking here will same a lot of time fixing, later.

Consider multiple daemons running at the same time.

It might be best to have the Game ID assigned by the MySQL database as an atomic operation, and have each daemon obtain a Game ID rather than try to make its own.

**Comment by Pentarctagon** [comment]
GAME_ID can be a regular INT if it's a concern.

The logic of having wesnothd create its own sequential IDs wouldn't be replaced by getting the ID from the database though, since the forum handler is optional, so adding a second way to get a sequential ID doesn't make much sense.

**Comment by ProditorMagnus** [comment]
I am unsure about scope of this now - is it supposed to save data of one restart cycle of wesnothd, or will it take all of 1.15, 1.16, ... that there ever will be?

> (it seemed unlikely that there would be 65k+ matches between restarts of wesnothd).

1.14 has been restarted, but game id is around 80k currently.

**Comment by Pentarctagon** [comment]
> I am unsure about scope of this now - is it supposed to save data of one restart cycle of wesnothd, or will it take all of 1.15, 1.16, ... that there ever will be?

My intent is for it to store the data indefinitely(or at least for a fairly long time), the same as the replay history which goes back to 1.6 (2010).  So this would keep data for 1.15, 1.16, 1.17, etc (and 2.0, 2.1, etc, if that comes to fruition).

> 1.14 has been restarted, but game id is around 80k currently.

Ah, so it is.  INT it is then.

**Comment by gfgtdf** [comment]
I think this idea is really nice and a first step to multiple useful features, in particular being able to reload games directly from the server or an 'ladder system' ( in case we decide that we want that). Or just as tool to get more data for debugging.

**Comment by GregoryLundberg** [comment]
So (in my dreams) Wesnoth is wildly popular and, to keep up with demand, we're running seven copies of Wesnothd for 1.15. Due to a power failure, all of them were restarted a few minutes ago. Now, we're getting complaints about something wrong. And the evidence is that we can check Game 62 from today. But there are 7 'Game 62' so someone needs to load each and see if it is the Game 62 we're looking for. Ah well, at least it's only seven we have to check, and not seven thousand.

**Comment by Pentarctagon** [comment]
If that's really a potential concern, then my question becomes: what would be a good way to retrieve the value of an AUTOINCREMENT column?

I know that Oracle databases have separate Sequences that can be used independently to create sequential values, and more generally that JDBC has the RETURN_GENERATED_KEYS constant that can be provided to a PreparedStatement to allow retrieving generated IDs without additional logic.

mysql seems to have the `mysql_insert_id()` C++ function or querying `LAST_INSERT_ID()`, both of which are separate operations.  I don't know enough about how wesnothd is implemented to know if separately calling either of those after the actual insert has taken place is safe or not.

**Comment by soliton-** [comment]
I think it would be useful to document somewhere what data you all want to save, were it comes from and what we hope to do with it. The idea would be to make it easier to verify that the intended purpose can actually be achieved and just to have some documentation of the DB model. Perhaps adding comments to the DB columns would be enough.

For example:
MAP_NAME; multiplayer settings from the host ([multiplayer] mp_scenario_name=); used to gather statistics(?)

Regarding that example I would rather save mp_scenario which is the ID. The name is presumably less stable and probably translated based on the host's language. Same for the era.

There are lots more scenario settings. Of course they're all in the replay data but if the purpose is to gather and analyze statistics then having some more easily available in the DB would be good. (e.g. fog, shroud, gold, timer)

For data like IS_HOST the question is what it's for since it can change over the course of a game. Perhaps the initial host that also provided the scenario data is good enough.

For REPLAY_NAME perhaps just save the filename and whoever wants to access the replay needs to configure a path prefix. At least for replays.wesnoth.org the path were replays are saved is not where they end up.

I would not block this PR for the TODO about deciding what sides win/lose, btw. We do have the surrender option now but otherwise wesnothd has no idea who won. That can be solved separately.

**Comment by Pentarctagon** [comment]
What about putting a `table_definitions.sql` file in the `src/server/` directory?  It would have comments/explanations in it, and anyone could then just take that and execute it to give them all the extra tables they'd need.

For MAP_NAME and ERA_NAME, that's a good point.

IS_HOST was just something I thought might be useful to record, but I have no strong opinion on whether it stays or goes.  Keeping it up to date would have the same problem I mentioned above as far as creating a primary key for the table though.

I'll update REPLAY_NAME.

For determining victory, there is actually something of the sort [already done](https://github.com/wesnoth/wesnoth/blob/master/src/playsingle_controller.cpp#L317-L323), though it's [currently ignored](https://github.com/wesnoth/wesnoth/blob/master/src/server/server.cpp#L1825-L1838).  My thought currently was to re-purpose that, and instead have each client send whether each local side has a leader alive, if it has units but no leader, or all units are dead, then look into adding something to surrender as well.  Given reloading, disconnects, etc, trying to conclusively determine a winning side for any single game feels like a bad idea.

As far as additional things to store, what I have here currently is just what I thought would be necessary or useful towards answering fairly simple questions like "what's the most played map (besides Isars)?" or "what's the most played faction?".  I've no objection to adding more information though, as long as disk space isn't a concern and it's information that's readily available already.

**Comment by CelticMinstrel** [comment]
If you're going to have an auto-increment column as a unique primary key, I'd say it should be `BIGINT`, but if the game ID is set by Wesnoth and Wesnoth only uses a 32-bit integer to store it, then `INT` should be fine; but you might want to have code to handle the case of overflow, just in case. Simply generating a new UUID if the game ID reaches `numeric_limits<uint32_t>::max()` should be sufficient, right?

Also, if one `GAME_ID` column is `INT`, then surely all `GAME_ID` columns should be `INT`? Aren't those supposed to be foreign key references to the main table? (Speaking of which... you're not using foreign key references?)

Finally, I'm not sure if the manpage is a good place for the table structure...? In any case, adding a .sql file would be useful.

**Comment by Pentarctagon** [comment]
My question about using an AUTO_INCREMENT column really comes down to whether wesnothd uses multiple threads for handling data sent by clients, since there's only one mysql connection created.  If it does, then the insert and the subsequent call to get the incremented ID might get interleaved and give the wrong value.

As far as foreign keys, I hadn't really thought about it.

I think we're going to run into a bunch of other issues if we really end up with over 4 billion rows, though.

I've fixed the INT/SMALLINT issue and added the sql file though.

**Comment by GregoryLundberg** [comment]
The number of MySQL connections does not matter. If parallel requests for an auto-increment column appear, MySQL handles it atommically and does not send out a duplicate (or there's a bug they need to fix).

It's a reasonably safe assumption that a big integer won't overflow in less than a day. So most systems couple the datetime with the auto-increment to uniquely identify an object. N.B.: This actually is not safe, but it's "safe enough" for most real-world applications.

Another option is to use a UUID. This, too, can be a problem.

Most UUID schemes are time dependent. For most, to avoid duplicated UUIDs, you run through a one-per-second rate-limited server to obtain a UUID.

Other schemes depend upon a random number generator and have a vanishingly small chance of assigning a dupllicated UUID. For these, you do not need to rate-limit but you need a database of all UUID every handed out. Simply check the new UUID does not already exist, and generate another if it does. Care needs to be taken to ensure there are no concurrency issues when faced with multiple, (near) simultaneous requests.

My real point with all this is, even if it takes a week or two of discussion, and a few more hours to implement, the time spent now is going to be far less than the time spend fixing the mess if we don't do it and that turns out to have been a bad idea.

**Comment by GregoryLundberg** [comment]
One more thought: making an auto-increment field a primary/unique key is a very bad idea.

**Comment by Pentarctagon** [comment]
My concern with the auto_increment ID isn't that it will generate duplicates, it's that if:
* thread 1 inserts
* thread 2 inserts
* thread 1 retrieves the id
* thread 2 retrieves the id
Then both thread 1 and thread 2 will get the id created by thread 2's insert.

As for using UUID(), the mysql documentation states that even for UUID_SHORT(), the limitation is: "You invoke UUID_SHORT() on average fewer than 16 million times per second between mysqld restarts".  But even if there was a collision just entirely at random, this is not sensitive or critical data, so I don't see the need to get particularly elaborate with the setup.

**Comment by soliton-** [comment]
Sending and receiving data used to be multi-threaded (currently wesnothd is completely single-threaded.) but the logic of handling that data is always single threaded.

**Comment by GregoryLundberg** [comment]
If thread 1 inserts, then thread 2 inserts, then thread 1 reads the auto-increment field from its currency, then thread 2 does the same, if they get the same value in the auto-increment field then submit a critical Zero-Day bug report to MySQL, you might check for an open CVE and get on assigned for them.

**Comment by ProditorMagnus** [comment]
If wesnothd processes incoming game data in multiple threads, then it will have more immediate issues of lobby game indexes not referring to games they are supposed to.

**Comment by soliton-** [comment]
> If wesnothd processes incoming game data in multiple threads, then it will have more immediate issues of lobby game indexes not referring to games they are supposed to.

It does not and never has.

**Comment by soliton-** [comment]
Nice work on the sql file. I agree that's better than inline in the man page. I would suggest to put it under `utils/mp-server/` though.

File: src/server/table_definitions.sql [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+--     username varchar(255) COLLATE utf8_bin NOT NULL DEFAULT '',
+--     user_password varchar(255) COLLATE utf8_bin NOT NULL DEFAULT '',
+--     user_email varchar(100) COLLATE utf8_bin NOT NULL DEFAULT '',
+--     PRIMARY KEY (user_id),
+--     KEY user_type (user_type)
+-- ) ENGINE=InnoDB AUTO_INCREMENT=50 DEFAULT CHARSET=utf8 COLLATE=utf8_bin;
+
+------
+
+-- table used to manage bans issued by wesnothd
```
Comment (soliton-) [inline src/server/table_definitions.sql]: The table is for forum bans and is only read from wesnothd. Bans issued from wesnothd go to `ban_save_file`

File: src/server/table_definitions.sql [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+DROP TABLE IF EXISTS extra;
+CREATE TABLE extra
+(
+    username varchar(100) NOT NULL,
+    user_lastvisit int(10) unsigned NOT NULL DEFAULT '0',
+    user_is_moderator tinyint(4) NOT NULL DEFAULT '0',
+    PRIMARY KEY (username)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;
+
+-- information about a single game instance
```
Comment (soliton-) [inline src/server/table_definitions.sql]: Perhaps a bit confusing to say game _instance_ when the columns with INSTANCE in their name refer to wesnothd.

**Comment by Pentarctagon** [comment]
Alright, it should be good now, I think, unless there are any strong opinions about using an AUTO_INCREMENT column or there's more pieces of data anyone would like added.

**Comment by CelticMinstrel** [comment]
> My concern with the auto_increment ID isn't that it will generate duplicates, it's that if:

Umm... databases are _built for concurrent access_. If what you describe were to happen, it would be (as @GregoryLundberg rightly noted) a critical bug in MySQL itself.

> One more thought: making an auto-increment field a primary/unique key is a very bad idea.

Why exactly is it a bad idea? Is it because eventually it'll overflow and create duplicates?

**Comment by GregoryLundberg** [comment]
Yes. Especially if you've used a smaller int.

We had a database, once, which had a lot of deletions as well as a lot of insertions. The original designer assumed that when the number overflowed, we'd be fine because the older record would be deleted. That worked fine .. for a couple/five years. Then, suddenly, mid-way through at some random point, it finally hit an old record which had not been deleted and things came crashing down.

I've also seen where a programmer, recognizing that a duplicate was impossible, went into a tight loop looking for the next available. And that worked fine. Except every now and then everything would stop as the server went into the loop jumping over large segments of existing records. This was made especially bad because the system would delete large swaths of older records, creating a sparsely populated key, with large clumps followed by large holes.

Don't get me wrong, I love auto-imcrement fields. But not as the sole, primary key field. I couple it with something else, like the date so, if it wraps, the primary key is still unique.

**Comment by soliton-** [comment]
@Pentarctagon Please remove the `src/playsingle_controller.cpp` change completely. :wink:

**Comment by Pentarctagon** [comment]
@CelticMinstrel What I mean/meant is that:
* wesnothd creates a single database connection when it starts up.
* mysql stores the most recently generated AUTO_INCREMENT value per connection.

Therefore, if wesnothd *were* multithreaded, it would mean that multiple threads were sharing that connection, which would mean that not having an atomic way to insert and retrieve that AUTO_INCREMENT value would be a problem.

**Comment by soliton-** [comment]
Perhaps remove the DROP TABLE statements from the sql file. I get that it's convenient for development puposes but normally people are not going to execute the SQL multiple times. If they do though currently it can be a quite unfortunate surprise.

Also I'd either comment or uncomment both the users and bans table since they either both come from phpbb or don't exist.

**Comment by GregoryLundberg** [comment]
@Pentarctagon You're confusing the key (which is atomic) and the read-out for the connection. Yes, you can get the last auto-increment value created for THIS CONNECTION but that is not necessarily THE last auto-increment value created, nor can it be used to guess the next value which will be generated.

The idea is that you insert a record, read the last AI value, store it locally with the record you just created, them move on. This will always give the value assigned (N) for the record, but does not mean the next insert will be assigned N+1 since other threads/connections from other machines/processes/users may have inserted in the meantime.

**Comment by Pentarctagon** [comment]
@soliton- Updated.

@GregoryLundberg I think maybe you're misunderstanding what my concern was.  I don't have any concerns whatsoever involving anything that happens on the mysql side of things.  My only concern was the potential for:
* Thread 1 inserts using `conn`, an ID of 5 is created.
* Thread 2 inserts using `conn`, an ID of 6 is created.
* Thread 1 retrieves the last AUTO_INCREMENT value from `conn`, gets 6 (instead of 5).
* Thread 2 retrieves the last AUTO_INCREMENT value from `conn`, gets 6.

**Comment by soliton-** [comment]
Cool, this is ready to merge from my point of view.  I'd be happy to do so if there are no further comments/concerns. Further tweaks as to what else we may want to save in the DB can be done incrementally.

**Comment by CelticMinstrel** [comment]
> Yes. Especially if you've used a smaller int.

Well yeah, that's why I said that if you _are_ doing it you should make the ID a `BIGINT` from the start.

> Therefore, if wesnothd _were_ multithreaded, it would mean that multiple threads were sharing that connection, which would mean that not having an atomic way to insert and retrieve that AUTO_INCREMENT value would be a problem.
>
>I think maybe you're misunderstanding what my concern was. I don't have any concerns whatsoever involving anything that happens on the mysql side of things. My only concern was the potential for:
>
> * Thread 1 inserts using `conn`, an ID of 5 is created.
> * Thread 2 inserts using `conn`, an ID of 6 is created.
> * Thread 1 retrieves the last AUTO_INCREMENT value from `conn`, gets 6 (instead of 5).
> * Thread 2 retrieves the last AUTO_INCREMENT value from `conn`, gets 6.

I'm pretty sure that would be impossible. If thread 2 were to try to use `conn` (by opening a transaction) while thread 1 were using `conn` (ie, with an open transaction), it would signal an error, because you already have a transaction open and there is only one transaction per session... though this is true for PostgreSQL, I can't be certain if it's true of MySQL...

More to the point though, if wesnothd _were_ made multithreaded, you would at that point have to work out this issue - either by pushing all database work to a single thread, or by having a separate database connection in each thread. For example, you might be able to declare `conn` as `thread_local`.

**Comment by Pentarctagon** [comment]
I guess it's a moot point, in any case.

**Comment by GregoryLundberg** [comment]
may be moot but worth pointing out that the race condition you're concerned about is not a MySQL issue. Its a pure-and-simple bug in the implementation of the threads. You either need a lock on the database, or, in this case, a lock around the insert/retrieve critical section. How you do it is up to you, but not doing it in your code is not an option.

**Comment by Pentarctagon** [comment]
Given I said I wasn't at all concerned with anything related to mysql itself... I guess we agree then?

**Comment by CelticMinstrel** [comment]
I'd also say that there's no point in worrying about possible race conditions when the program is single-threaded.

**Pull request merged**